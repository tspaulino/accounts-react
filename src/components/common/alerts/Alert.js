import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Message, Icon } from 'semantic-ui-react'

class Alert extends Component {
  componentWillMount() {
    const { revokeAlert, id } = this.props

    // Schedule alerts to be
    // removed after 7 seconds
    this.timer = setTimeout(() => revokeAlert(id), 7000)
  }

  shouldComponentUpdate() {
    return false
  }

  componentWillUnmount() {
    clearTimeout(this.timer)
  }

  render() {
    const { message, type } = this.props
    const props = { [type]: true }
    return (
      <Message {...props}>
        <Icon name="warning" />
        {message}
      </Message>
    )
  }
}

Alert.propTypes = {
  revokeAlert: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired
}

export default Alert
