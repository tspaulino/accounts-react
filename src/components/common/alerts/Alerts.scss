.alerts {
  position: absolute;
  top: 20px;
  width: 400px;
  margin: 0 auto;
  left: 0;
  right: 0;
  z-index: 1000;
}
