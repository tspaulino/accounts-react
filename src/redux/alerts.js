import uniqueId from 'lodash/uniqueId'
import isString from 'lodash/isString'
import find from 'lodash/find'
import without from 'lodash/without'

import createReducer from '../utils/createReducer'

// Actions
export const EMIT_ALERT = 'EMIT_ALERT'
export const REVOKE_ALERT = 'REVOKE_ALERT'

let alerts = {
  EMPTY_FIELDS: 'Please solve the errors bellow',
  GENERAL_ERROR: 'An error ocurred, please try again later'
}

const getAlert = (codeId, type) => {
  const code = alerts[codeId] ? codeId : 'GENERAL_ERROR'

  return {
    id: uniqueId('alert_'),
    type: isString(type) ? type : 'error',
    content: alerts[code],
    code,
  }
}

export const addAlerts = (newAlerts = {}) => {
  alerts = Object.assign({}, alerts, newAlerts)
}

// Action creators
export const emitAlert = (code, type) => ({ type: EMIT_ALERT, payload: getAlert(code, type) })
export const revokeAlert = id => ({ type: REVOKE_ALERT, payload: { id } })

// Initial state
const initialState = {
  items: []
}

// Reducer
const actionHandlers = {
  [EMIT_ALERT]: (state, { payload }) => ({ items: state.items.concat(payload) }),
  [REVOKE_ALERT]: (state, { payload }) => {
    const row = find(state.items, { id: payload.id })
    const items = (row) ? without(state.items, row) : state.items

    return { items }
  }
}

export default createReducer(initialState, actionHandlers)
