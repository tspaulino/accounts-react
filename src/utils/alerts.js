/**
 * Alerts util
 *
 * @copyright Copyright (c) 2017, hugw.io
 * @author Hugo W - me@hugw.io
 * @license MIT
 */

import uniqueId from 'lodash/uniqueId'
import isString from 'lodash/isString'
import pick from 'lodash/pick'
import isEmpty from 'lodash/isEmpty'


const alerts = {
  fieldErrors: { message: 'Please solve the errors bellow', type: 'error' },
  generalError: { message: 'An error ocurred, please try again later', type: 'error' },
  notAuthenticated: { message: 'Please sign in to continue', type: 'error' }
}

export default (alert) => {
  const id = uniqueId('alert')

  if (!alert || isEmpty(alert) || isString(alert)) {
    return {
      id,
      ...(alerts[alert] || alerts.generalError)
    }
  }

  return { id, type: 'error', ...pick(alert, ['message', 'type', 'id']) }
}
