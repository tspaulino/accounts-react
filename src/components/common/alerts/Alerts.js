import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import PropTypes from 'prop-types'

import { revokeAlert } from '../../../redux/alerts'
import Alert from './Alert'
import './Alerts.scss'

export const Alerts = ({ items, actions }) => (
  <div className="alerts">
    {items.map(item => <Alert key={item.id} {...item} {...actions} />)}
  </div>
)

Alerts.propTypes = {
  actions: PropTypes.shape({
    revokeAlert: PropTypes.func.isRequired,
  }).isRequired,
  items: PropTypes.arrayOf(PropTypes.object).isRequired
}

const mapStateToProps = state => ({ items: state.alerts.items })
const mapDispatchToProps = dispatch => ({ actions: bindActionCreators({ revokeAlert }, dispatch) })

export default connect(mapStateToProps, mapDispatchToProps)(Alerts)
