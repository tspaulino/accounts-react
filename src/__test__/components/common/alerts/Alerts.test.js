import React from 'react'
import { shallow } from 'enzyme'
import sinon from 'sinon'

import Alert from '../../../../components/common/alerts/Alert'
import { Alerts } from '../../../../components/common/alerts/Alerts'

const alerts = [{
  id: 'id1',
  type: 'error',
  message: 'Some error 1'
},
{
  id: 'id2',
  type: 'error',
  message: 'Some error 2'
},
{
  id: 'id3',
  type: 'error',
  message: 'Some error 3'
}]

const props = {
  items: alerts,
  actions: {
    revokeAlert: sinon.spy()
  }
}

describe('<Alert />', () => {
  it('should render component', () => {
    const wrapper = shallow(<Alerts {...props} />)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.is('.alerts')).toBeTruthy()
    expect(wrapper.find(Alert).length).toEqual(3)
  })
})
