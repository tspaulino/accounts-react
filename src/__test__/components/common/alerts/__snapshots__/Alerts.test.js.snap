// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> should render component 1`] = `
<div
  className="alerts"
>
  <Alert
    id="id1"
    key="id1"
    message="Some error 1"
    revokeAlert={[Function]}
    type="error"
  />
  <Alert
    id="id2"
    key="id2"
    message="Some error 2"
    revokeAlert={[Function]}
    type="error"
  />
  <Alert
    id="id3"
    key="id3"
    message="Some error 3"
    revokeAlert={[Function]}
    type="error"
  />
</div>
`;
